src/common/signal.cpp

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "mir/signal.h"
+
+mir::Signal::Signal()
+  : flag{false}
+{
+}
+
+void mir::Signal::raise()
+{
+    flag = true;
+    flag.notify_all();
+}
+
+void mir::Signal::wait()
+{
+    flag.wait(false);
+    flag = false;
```

Comment:
Isn't this racy? `raise()` can be called between these lines.

Reply:
It *is* racy, but the race is benign for our purposes, and is documented by 
> ```* Two unsynchronised calls to `raise()` may result in either one```
> ```* or two calls to `wait()` being unblocked, depending on timing.```

This is also not a race that's fixed by using a mutex + condition variable; two unsynchronised `raise()`s will still result in either one or two `wait()`s being unblocked, depending on whether both `raise()`s grab the mutex before the `wait()` does.

Reply:
Ack